src/atlas/interpolation/element/Quad2D.cc

```diff
@@ -0,0 +1,227 @@
+/*
+ * (C) Crown Copyright 2021 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#include <cmath>
+#include <iostream>
+
+#include "atlas/interpolation/Vector2D.h"
+#include "atlas/interpolation/element/Quad2D.h"
+#include "atlas/interpolation/element/Triag2D.h"
+#include "atlas/runtime/Log.h"
+
+namespace atlas {
+namespace interpolation {
+namespace element {
+
+//----------------------------------------------------------------------------------------------------------------------
+
+method::Intersect Quad2D::intersects(const PointXY& r, double edgeEpsilon, double epsilon) const {
+    method::Intersect isect;  // intersection is false
+
+    /* Split quadrilateral into two triangles, points are labelled counter-clockwise.
+     *   v01-----------v11
+     *    /  " .        `.
+     *   /         " .    `.
+     *  /                " .`.
+     * v00-------------------v10
+     *
+     */
+
+    Triag2D T013(v00, v10, v01);
+    isect = T013.intersects(r, edgeEpsilon, epsilon);
+    if (isect) {
+        return isect;
+    }
+
+    Triag2D T231(v11, v01, v10);
+    isect = T231.intersects(r, edgeEpsilon, epsilon);
+    if (isect) {
+        isect.u = 1 - isect.u;
+        isect.v = 1 - isect.v;
+        return isect;
+    }
+
+    return isect.fail();
+}
+
+method::Intersect Quad2D::localRemap(const PointXY& p, double edgeEpsilon, double epsilon) const {
+    method::Intersect isect;
+
+    // work out if point is within the polygon
+    if (!inQuadrilateral({p.x(), p.y()}))
+        return isect.fail();
+
+    // calculate local coordinates via remapping technique
+    // if that fails fall back to finite-element scheme.
+    //
+    // ax**2 + bx + x = 0
+    auto solve_quadratic = [](const double a, const double b, const double c) {
+        double det = b * b - 4 * a * c;
+        Roots roots;
+        if (det >= 0.) {
+            roots.a = (-b + sqrt(det)) / (2 * a);
+            roots.b = (-b - sqrt(det)) / (2 * a);
```

Comment:
Performance, and promote integers in expressions to double.
```suggestion
            double inv_two_a = 1. / (2. * a);
            double sqrt_det = std::sqrt(det);
            roots.a = (-b + sqrt_det) * inv_two_a;
            roots.b = (-b - sqrt_det) * inv_two_a;
```